corehq/util/doc_processor/couch.py

```diff
@@ -0,0 +1,135 @@
+from couchdbkit import ResourceNotFound
+
+from corehq.util.couch_helpers import PaginatedViewArgsProvider
+from corehq.util.doc_processor.interface import DocumentProvider, ProcessorProgressLogger
+from corehq.util.pagination import ResumableFunctionIterator
+
+
+class ResumableDocsByTypeArgsProvider(PaginatedViewArgsProvider):
+    def __init__(self, initial_view_kwargs, doc_types):
+        super(ResumableDocsByTypeArgsProvider, self).__init__(initial_view_kwargs)
+        self.doc_types = doc_types
+
+    def get_next_args(self, result, *last_args, **last_view_kwargs):
+        try:
+            return super(ResumableDocsByTypeArgsProvider, self).get_next_args(
+                result, *last_args, **last_view_kwargs
+            )
+        except StopIteration:
+            last_doc_type = last_view_kwargs["startkey"][0]
+            # skip doc types already processed
+            index = self.doc_types.index(last_doc_type) + 1
+            self.doc_types = self.doc_types[index:]
+            try:
+                next_doc_type = self.doc_types[0]
+            except IndexError:
+                raise StopIteration
+            last_view_kwargs.pop('skip', None)
+            last_view_kwargs.pop("startkey_docid", None)
+            last_view_kwargs['startkey'] = [next_doc_type]
+            last_view_kwargs['endkey'] = [next_doc_type, {}]
+        return last_args, last_view_kwargs
+
+
+def ResumableDocsByTypeIterator(db, doc_types, iteration_key, chunk_size=100, view_event_handler=None):
+    """Perform one-time resumable iteration over documents by type
+
+    Iteration can be efficiently stopped and resumed. The iteration may
+    omit documents that are added after the iteration begins or resumes
+    and may include deleted documents.
+
+    :param db: Couchdb database.
+    :param doc_types: A list of doc type names to iterate on.
+    :param iteration_key: A unique key identifying the iteration. This
+    key will be used in combination with `doc_types` to maintain state
+    about an iteration that is in progress. The state will be maintained
+    indefinitely unless it is removed with `discard_state()`.
+    :param chunk_size: Number of documents to yield before updating the
+    iteration checkpoint. In the worst case about this many documents
+    that were previously yielded may be yielded again if the iteration
+    is stopped and later resumed.
+    """
+
+    def data_function(**view_kwargs):
+        return db.view('all_docs/by_doc_type', **view_kwargs)
+
+    doc_types = sorted(doc_types)
```

Comment:
If you decide to change it to use the natural ordering of whatever `doc_types` sequence is passed in to allow prioritization, it should be documented that the sequence order must never change while the iteration is in progress since that could cause some doc types to be omitted completely if the iteration was stopped and then resumed with a different doc type ordering. Might want to also check that they didn't pass a dict or set since the iteration order of those types can change for [different interpreter runs](https://lwn.net/Articles/491788/).